bluemira/geometry/varied_offset.py

```diff
@@ -0,0 +1,165 @@
+# bluemira is an integrated inter-disciplinary design tool for future fusion
+# reactors. It incorporates several modules, some of which rely on other
+# codes, to carry out a range of typical conceptual fusion reactor design
+# activities.
+#
+# Copyright (C) 2022 M. Coleman, J. Cook, F. Franza, I.A. Maione, S. McIntosh, J. Morris,
+#                    D. Short
+#
+# bluemira is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# bluemira is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with bluemira; if not, see <https://www.gnu.org/licenses/>.
+"""
+Module containing functions to generate variable offset curves
+"""
+
+import numpy as np
+
+from bluemira.geometry.tools import find_clockwise_angle_2d, make_bspline
+from bluemira.geometry.wire import BluemiraWire
+
+
+def varied_offset(
+    wire: BluemiraWire,
+    min_offset: float,
+    max_offset: float,
+    min_offset_angle: float,
+    max_offset_angle: float,
+    num_points: int = 200,
+) -> BluemiraWire:
+    """
+    Create a new wire that offsets the given wire using a variable
+    offset in the xz plane.
+
+    All angles are measured from the negative x-direction (9 o'clock),
+    centred at the center of mass of the wire.
+    The offset will be 'min_offset' between the negative x-direction
+    and 'min_offset_angle'. Between 'max_offset_angle' and
+    the positive x-direction the offset will be 'max_offset'. Between
+    those angles, the offset will linearly transition between the min
+    and max.
+
+    Parameters
+    ----------
+    wire: BluemiraWire
+        The wire to create the offset from. This should be convex in
+        order to get a sensible, non-intersecting, offset.
+    min_offset: float
+        The size of the minimum offset.
+    max_offset: float
+        The size of the maximum offset.
+    min_offset_angle: float
+        The angle at which the variable offset should begin, in degrees.
+    max_offset_angle: float
+        The angle at which the variable offset should end, in degrees.
+    num_points: int
+        The number of points to use in the discretization of the input
+        wire.
+
+    Returns
+    -------
+    offset_wire: BluemiraWire
+        New wire at a variable offset to the input.
+    """
```

Comment:
check for acceptable input wires:
* closed
* planar  (x-z) 
* convex (??)

And raise GeometryError or warn/modify if not.

Reply:
Just for fun, I tried it with a straight line. Amazingly, no errors, but probably not what one mine expect :D:
![image](https://user-images.githubusercontent.com/26097289/160064142-5e9ea6a5-6e94-41af-af13-110377d624e1.png)

Reply:
😆 I'll add a check that the wire's closed